staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -144,7 +144,14 @@ func (c *controller) Run(stopCh <-chan struct{}) {
 	c.reflectorMutex.Unlock()
 
 	var wg wait.Group
-	defer wg.Wait()
+	// panics should bypass the wg.Wait and propagate up in this defer func
+	// because the process should crash
+	defer func() {
+		if r := recover(); r != nil {
+			panic(r)
+		}
+		wg.Wait()
+	}()
```

Comment:
Is it possible to write a unit test for this that actually creates a panic? The test case could then recover() from the panic in order to verify that it is being correctly propagated.

This change makes sense (check for panics before waiting). But given the confusing nature of defer funds, hard to be 100% certAin 😄

I’m also slightly concerned about the behavioural change here. Could this cause existing controllers to begin crashing where they previously would not? If so, is there any way we can avoid the ‘stuck’ defer call without changing this behaviour?

Reply:
Also worth looking at staging/src/k8s.io/client-go/tools/cache/shared_informer.go. https://github.com/kubernetes/kubernetes/pull/93679/files addresses a similar problem in that code. Oddly this fix may be more appropriate for dealing with that instance as it maintains relative code ordering.

Reply:
I don't think `staging/src/k8s.io/client-go/tools/cache/shared_informer.go` has the same problem. Because the event handler world be called in another goroutine, the panic will propagate and let the process exit as we wish .

In the shared_informer, the event handler will called in `wg.StartWithChannel(processorStopCh, s.processor.run)`.

`wg.StartWithChannel` will create a new goroutine and the panic from `s.processor.run` would not stuck.

And actually, `s.controller.Run(stopCh)` in shared_informer just run Reflector,  wouldn't run the handler which may call a panic. It's not completely similar to the informer controller.

Reply:
> Is it possible to write a unit test for this that actually creates a panic? The test case could then recover() from the panic in order to verify that it is being correctly propagated.

I had written an unit test. What do you think about it?